File: VB/App_Code/TestHandler.vb
Option Infer On

Imports System
Imports System.Web
Imports System.Web.SessionState

Public Class TestHandler
    Implements IHttpHandler, IRequiresSessionState

    Private Function GetProgress(ByVal session As HttpSessionState) As String
        If session Is Nothing OrElse Not (TypeOf session("_Operation") Is Operation) Then
            Return "NaN"
        End If
        Dim progress = (TryCast(session("_Operation"), Operation)).Progress
        Return progress.ToString()
    End Function

    Public ReadOnly Property IsReusable() As Boolean Implements IHttpHandler.IsReusable
        Get
            Return False
        End Get
    End Property

    Public Sub ProcessRequest(ByVal context As HttpContext) Implements IHttpHandler.ProcessRequest
        Dim response = GetProgress(context.Session)
        context.Response.CacheControl = "No-cache"
        context.Response.Write(response)
    End Sub
End Class

File: CS/App_Code/TestHandler.cs
using System;
using System.Web;
using System.Web.SessionState;

public class TestHandler : IHttpHandler, IRequiresSessionState {
    
    String GetProgress(HttpSessionState session) {
        if(session == null || !(session["_Operation"] is Operation))
            return "NaN";
        var progress = (session["_Operation"] as Operation).Progress;
        return progress.ToString();
    }

    public Boolean IsReusable {
        get { return false; }
    }

    public void ProcessRequest(HttpContext context) {
        var response = GetProgress(context.Session);
        context.Response.CacheControl = "No-cache";
        context.Response.Write(response);
    }
}